cs-mail/src/main/java/io/cloudslang/content/mail/services/SendMailService.java

```diff
@@ -0,0 +1,312 @@
+
+            if(!StringUtils.isEmpty(input.getUser())) {
+                transport = session.getTransport(SmtpPropNames.SMTP);
+                transport.connect(input.getHostname(), input.getPort(), input.getUser(), input.getPassword());
+                transport.sendMessage(msg, msg.getAllRecipients());
+            } else {
+                Transport.send(msg);
+            }
+            result.put(OutputNames.RETURN_RESULT, Constants.MAIL_WAS_SENT);
+            result.put(OutputNames.RETURN_CODE, ReturnCodes.SUCCESS);
+        } finally {
+            if (null != transport) {
+                transport.close();
+            }
+        }
+        return result;
+    }
+
+    private void processHTMLBodyWithBASE64Images(MimeMultipart multipart) throws ParserException,
+            MessagingException, NoSuchAlgorithmException, SMIMEException, java.security.NoSuchProviderException {
+        if (null != input.getBody() && input.getBody().contains(Encodings.BASE64)) {
+            Parser parser = new Parser(input.getBody());
+            NodeList nodeList = parser.parse(null);
+            HtmlImageNodeVisitor htmlImageNodeVisitor = new HtmlImageNodeVisitor();
+            nodeList.visitAllNodesWith(htmlImageNodeVisitor);
+            input.setBody(nodeList.toHtml());
+
+            addAllBase64ImagesToMimeMultipart(multipart, htmlImageNodeVisitor.getBase64Images());
+        }
+    }
+
+    private void addAllBase64ImagesToMimeMultipart(MimeMultipart multipart, Map<String, String> base64ImagesMap)
+            throws MessagingException, NoSuchAlgorithmException, NoSuchProviderException, SMIMEException {
+        for (String contentId : base64ImagesMap.keySet()) {
+            MimeBodyPart imagePart = getImageMimeBodyPart(base64ImagesMap, contentId);
+            imagePart = encryptMimeBodyPart(imagePart);
+            multipart.addBodyPart(imagePart);
+        }
+    }
+
+    private MimeBodyPart getImageMimeBodyPart(Map<String, String> base64ImagesMap, String contentId)
+            throws MessagingException {
+        MimeBodyPart imagePart = new MimeBodyPart();
+        imagePart.setContentID(contentId);
+        imagePart.setHeader(Encodings.CONTENT_TRANSFER_ENCODING, Encodings.BASE64);
+        imagePart.setDataHandler(new DataHandler(Base64.decode(base64ImagesMap.get(contentId)), MimeTypes.IMAGE_PNG));
+        return imagePart;
+    }
+
+    private MimeBodyPart encryptMimeBodyPart(MimeBodyPart mimeBodyPart) throws NoSuchAlgorithmException,
+            NoSuchProviderException, SMIMEException {
+        if (input.isEncryptedMessage()) {
+            mimeBodyPart = gen.generate(mimeBodyPart, input.getEncryptionAlgorithm(), SecurityConstants.BOUNCY_CASTLE_PROVIDER);
+        }
+        return mimeBodyPart;
+    }
+
+    private void addEncryptionSettings() throws Exception {
+        URL keystoreUrl = new URL(input.getEncryptionKeystore());
+        InputStream publicKeystoreInputStream = keystoreUrl.openStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *publicKeystoreInputStream*. The resource is referenced in statements at the following line: 251. The resource closure statement is at line: 253. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by KeyStore.getInstance. Either a) close *publicKeystoreInputStream* in a try-finally block or b) close the resource by declaring *publicKeystoreInputStream* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).